FHEM/00_SIGNALduino.pm

```diff
@@ -2031,100 +2031,124 @@ SIGNALduino_Parse_MS($$$$%)
-			#Check calculated max length
-			$valid = $valid && $ProtocolListSIGNALduino{$id}{length_max} >= $bit_length if (exists $ProtocolListSIGNALduino{$id}{length_max});
-
-			Debug "expecting $bit_length bits in signal" if ($debug);
-			next if (!$valid) ;
+			if ($ProtocolListSIGNALduino{$id}{clockabs} > 0) {
+				if (!SIGNALduino_inTol($ProtocolListSIGNALduino{$id}{clockabs},$clockabs,$clockabs*0.30)) {
+					Debug "protocClock=$ProtocolListSIGNALduino{$id}{clockabs}, msgClock=$clockabs is not in tol=" . $clockabs*0.30 if ($debug);  
+					next;
+				} elsif ($debug) {
+					Debug "protocClock=$ProtocolListSIGNALduino{$id}{clockabs}, msgClock=$clockabs is in tol=" . $clockabs*0.30;
+				}
+			}
 
 			#Debug Dumper(@{$ProtocolListSIGNALduino{$id}{sync}});
 			Debug "Searching in patternList: ".Dumper(\%patternList) if($debug);
-
 			Debug "searching sync: @{$ProtocolListSIGNALduino{$id}{sync}}[0] @{$ProtocolListSIGNALduino{$id}{sync}}[1]" if($debug); # z.B. [1, -18] 
 			#$valid = $valid && SIGNALduino_inTol($patternList{$clockidx}, @{$ProtocolListSIGNALduino{$id}{sync}}[0], 3); #sync in tolerance
 			#$valid = $valid && SIGNALduino_inTol($patternList{$syncidx}, @{$ProtocolListSIGNALduino{$id}{sync}}[1], 3); #sync in tolerance
 			
 			my $pstr;
 			my %patternLookupHash=();
+			my %endPatternLookupHash=();
 
 			$valid = $valid && ($pstr=SIGNALduino_PatternExists($hash,\@{$ProtocolListSIGNALduino{$id}{sync}},\%patternList,\$rawData)) >=0;
 			Debug "Found matched sync with indexes: ($pstr)" if ($debug && $valid);
 			$patternLookupHash{$pstr}="" if ($valid); ## Append Sync to our lookuptable
+			$endPatternLookupHash{substr($pstr,0,length($pstr)-1)}="" if ($valid); ## Append Sync to our lookuptable
 			my $syncstr=$pstr; # Store for later start search
 
 			Debug "sync not found " if (!$valid && $debug); # z.B. [1, -18] 
 
 			next if (!$valid) ;
-
+			
+			my $message_start =index($rawData,$syncstr)+length($syncstr);
+			my $signal_width= @{$ProtocolListSIGNALduino{$id}{one}};
+			my $bit_length = ($signal_length-$message_start) / $signal_width;
+			Debug "expecting $bit_length bits in signal" if ($debug);
+			
+			#Check calculated min length
+			if (exists($ProtocolListSIGNALduino{$id}{length_min}) && $ProtocolListSIGNALduino{$id}{length_min} > $bit_length) {
+				Debug "bit_length=$bit_length to short" if ($debug);
+				next;
+			}
+			
 			$valid = $valid && ($pstr=SIGNALduino_PatternExists($hash,\@{$ProtocolListSIGNALduino{$id}{one}},\%patternList,\$rawData)) >=0;
 			Debug "Found matched one with indexes: ($pstr)" if ($debug && $valid);
-			$patternLookupHash{$pstr}="1" if ($valid); ## Append Sync to our lookuptable
+			$patternLookupHash{$pstr}="1" if ($valid); ## Append one to our lookuptable
+			$endPatternLookupHash{substr($pstr,0,length($pstr)-1)}="1" if ($valid); ## Append one to our endbit lookuptable
```

Comment:
Bei einer $signal_width von 1 passt dies nicht so richtig, 
ich hatte es so gemacht:
```
$endPatternLookupHash{substr($pstr,0,$signal_width-1)}="1" if ($valid && exists($ProtocolListSIGNALduino{$id}{reconstructBit}));
```

Reply:
In beiden Fällen gibt das 0.
Wenn ein Bit nur aus einem Puls besteht, können wir nicht rekonstruieren oder siehst Du das anders?

Reply:
Da bei einer $signal_width ein rekonstruieren nicht möglich ist, hatte ich bei meinem pullrequest ein  `if (exists($ProtocolListSIGNALduino{$id}{reconstructBit})) ` drin,
dies wurde durch #491 wieder überschrieben

Reply:
Ich verstehe nicht ganz was signal_width mit reconstructBit zu tun hat? Wenn 'reconstructBit' gesetzt ist, aber eine signal_width von 1 berechnet wurde haben wir doch das gleiche Verhalten.


Was genau passt denn nicht? Ich hab ein kleines Testprogramm erstellt:
http://tpcg.io/sVJAb6

Da kommt es zu keinem Fehler.

Reply:
Das reconstructBit sollte nur bei den Protokollen definiert werden wo es auch benötigt wird und sinnvoll ist.
Meinst Du jemand würde das reconstructBit in die Protokolldefinition eintragen ohne vorher nachzudenken ob es überhaupt sinn macht?
Ein `$ProtocolListSIGNALduino{$id}{} = "1"` finde ich unnötig und unsauber, kannst Du ausschließen, daß diese unsauberkeit nicht mal in einer zukünftigen Perl Version ein warning erzeugt?

Reply:
genau genommen steht da $endPatternLookupHash{""} und damit haben wir einen String ohne Inhalt. Und ja das wird derzeit durch Perl unterstützt.


Ob man beim Setzen von `reconstructBit` beeinflussen kann, ob die Länge der Signale für 0 und 1 identisch und lang genug ist, wage ich aus Erfahrung bei MU Protokollen zu bezweifeln.

Um all die Fälle abzufangen wäre folgender Code notwendig:
```
if (exists($ProtocolListSIGNALduino{$id}{reconstructBit}) && lenght($pstr) > 1)
{
     chop($pstr);
     $endPatternLookupHash{$pstr}="1"
}
```
Da dachte ich mir halt, dass man sich die Abfragen sparen kann, da es keine Auwirkung hat wenn der $pstr zu klein wird.

Reply:
ich habe nochmal darüber nachgedacht.
Auf eine Abfrage bei $signal_width = 1 können wir verzichten. Eine ungerade $signal_width macht doch keinen Sinn, wie soll dies decodiert werden?
Daß die signal_width für one und zero unterschiedlich sein könnte ist auch nur eine fiktive Annahme, wie soll so was decodiert werden, wenn z.B. zero eine signal_width von 2 und one eine signal_width von 4 hat.

Ein length($pstr) halte ich für unnötig da dies schon in $signal_width steht.